pymeasure/instruments/instrument.py

```diff
@@ -293,23 +312,42 @@ def fset(self, value):
     def clear(self):
         """ Clears the instrument status byte
         """
-        self.write("*CLS")
+        if self.SCPI:
+            self.write("*CLS")
+        else:
+            raise NotImplementedError("Non SCPI instruments require implementation in subclasses")
 
     # TODO: Determine case basis for the addition of this method
     def reset(self):
         """ Resets the instrument. """
-        self.write("*RST")
+        if self.SCPI:
+            self.write("*RST")
+        else:
+            raise NotImplementedError("Non SCPI instruments require implementation in subclasses")
 
     def shutdown(self):
         """Brings the instrument to a safe and stable state"""
         self.isShutdown = True
         log.info("Shutting down %s" % self.name)
 
     def check_errors(self):
```

Comment:
Great, thank you for adding this generic implementation!
Could you please also remove all the instrument-specific `check_errors` that are now being made redundant by this change?

Reply:
Thanks for the review, I had a second thought: I really don't like all that `if self.SCPI` that are against proper use of classes. What about having having `Instrument` class as a base class for SCPI and non-SCPI instruments and then having a proper subclass `SCPIInstrument` which implements all the mandatory SCPI commands ?
That is essentially moving all the following properties to subclass `SCPIInstrument`:

- `complete`
- `status`
- `options`
- `id`

and changing methods: `clear`, `reset`, `shutdown`, `check_errors` with 'Subclass should implement this' exception and make the SCPI implementation in `SCPIInstrument`. Also parameter `includeSCPI` should be removed from `__init__`.
After this, all the implemented instruments that are SCPI compliant should be changed to inherits from `SCPIInstrument` and remove the possible parameter `includeSCPI` and `check_errors` implementation. All the other should be left untouched.

Reply:
Fair point. I'm a bit hesitant to make the inheritance structure more complicated, but in this case it's probably straightforward enough - `SCPIInstrument` isa `Instrument` is a pretty clear-cut relationship. Also, the documentation needs to be updated.

Normally, I'd try nowadays to add this additional behaviour as a Mixin (to favour composition over inheritance), e.g. like [here](https://stackoverflow.com/a/39471532), but it makes the correct inheritance formulation more fragile for our instrument contributors (correct order, correct use of `super()`), and I currently don't expect that we need more classes in the inheritance tree.

However, *if* we go towards these "reuseable, common-functionality" instrument classes (SignalGenerator and such) like you proposed in #416, this will get more complicated to deal with cleanly. Suddenly there is another axis to consider - to we make `SignalGenerator` *and* `SCPISignalGenerator` (to me this feels quite unnecessarily complicated)? 
As the scpi command set can be considered a feature/trait, maybe it would be better to use an `SCPIMixin`, after all, and any concrete implementation would then be `class KeithleyXYZ(SCPIMixin, SignalGenerator)` or `class RSXYZ(SignalGenerator)`? We then have to educate contributors on the importance of `super()`, order of base classes/MRO, and `*args, **kwargs` pass-through in initialisers. Probably this can best be sorted out with good Instrument templates/documentation.

Think about it!

Reply:
It sounds like a very straightforward implementation, and I indeed can see how it is a bit neater than the using the `includeSCPI` option. For the sake of the contributors I'm not sure if I'm a big fan of the Mixin approach (I understand how it would be nicer, but I also think that in order to keep contributing instruments as accessible as possible it would be nice if a specific instrument only has to inherit a single superclass).

> to we make SignalGenerator and SCPISignalGenerator (to me this feels quite unnecessarily complicated)

I'm not so sure if that would actually be required: are there actually any shared commands between non-SCPI signal generators, as I'd think that all/most shared commands have the stem from the SCPI standard (but maybe I'm wrong though).

Reply:
If I understand correctly what you are saying, you argue that, should we implement #416, we base all the new instrument "categories" on `SCPIInstrument` (naming either `SCPISignalGenerator` or just the simpler `SignalGenerator` might be preferable), and any non-scpi devices (either just not complying to one or more standard commands, or e.g. thorlabs devices) are based solely on `Instrument`. Correct?
This means we lose the opportunity to classify those non-scpi devices, as they are `Instruments`, but that might be an acceptable drawback.

Reply:
@bilderbuchi, thanks for highlighting these very sharp points which I did not see before your comment. I understand and I share your concerns regarding having multiple dimension and multiple inheritance with #416, my gut feelings is that we could implement the `SCPIInstrument` class and then adding instrument categories either as a Mixin or directly inheriting from `SCPIInstrument`. I see some tendency with T&M manufacturers to make more compact multi-function devices and in such case, I think, a Mixin approach would be good but with more complicated inheritance structure which, at the end of the day, probably reflect the fact that instrument is more complex.

I think some kind of decision needs to be made: shall I try to implement `SCPIInstrument` as described in https://github.com/pymeasure/pymeasure/pull/453#discussion_r651306658 ? Or just implement the changes requested in this PR and leave the `SCPIInstrument` changes for another PR ?
@CasperSchippers , @bilderbuchi any preference/suggestion from your side ?

Reply:
Another point to consider: 
One thing we (I think) want from the instrument categories (#416) is instrument API consistency. 
With the possible exception of the standard SCPI commands, this should be independent from SCPI-conformance or not. So, we will need some kind of mechanism that prescribes the available commands. If that should happen via Mixin (SCPI or not), or via inheritance and subsequent overriding of properties in the subclass, is not clear at this point I would say. This needs more discussion/thinking IMO.

I think it's safer to 
* wait with the `SCPIInstrument` implementation (the current solution is already in place and not so ugly)
* implement the requested changes to get this PR merged
* In #416, link to this discussion/the most valuable points so that they are not lost
* and add the `SCPIInstrument` feature/improvement (mixin or subclass or ...) to the scope of #416.

Let's see if @CasperSchippers agrees :-)

Reply:
I think I agree with that, @bilderbuchi. For now indeed I'd finish this and merge the PR and add the SCPIInstrument implementation to the scope of #416.

> If I understand correctly what you are saying..... 

Yes that was indeed what I meant (I was mainly thinking from the point of view of shared commands), though I agree with your point regarding API consistency (although I'm not yet sure how to best implement it at this moment).